### ISSUE TITLE
Load error with Cygwin and non-Cygwin Git

### ISSUE THREAD

**Issue by CyberShadow** [magit/magit#2503]
Trying to load magit on a system which has Cygwin and `git` in its path, but not using Cygwin's git (e.g. Git for WIndows), results in an error with the following stack trace:

```
Debugger entered--Lisp error: (file-error "Searching for program" "no such file or directory" "C:\\cygwin\\libexec\\git-core\\git")
  call-process("C:\\cygwin\\libexec\\git-core\\git" nil #<buffer  *temp*> nil "-c" "alias.echo=!echo" "echo" "x{0}")
  apply(call-process "C:\\cygwin\\libexec\\git-core\\git" nil #<buffer  *temp*> nil ("-c" "alias.echo=!echo" "echo" "x{0}"))
  process-file("C:\\cygwin\\libexec\\git-core\\git" nil #<buffer  *temp*> nil "-c" "alias.echo=!echo" "echo" "x{0}")
  (let ((process-environment (append magit-git-environment process-environment))) (process-file magit-git-executable nil (current-buffer) nil "-c" "alias.echo=!echo" "echo" "x{0}"))
  (progn (let ((process-environment (append magit-git-environment process-environment))) (process-file magit-git-executable nil (current-buffer) nil "-c" "alias.echo=!echo" "echo" "x{0}")) (buffer-string))
  (unwind-protect (progn (let ((process-environment (append magit-git-environment process-environment))) (process-file magit-git-executable nil (current-buffer) nil "-c" "alias.echo=!echo" "echo" "x{0}")) (buffer-string)) (and (buffer-name temp-buffer) (kill-buffer temp-buffer)))
  (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn (let ((process-environment (append magit-git-environment process-environment))) (process-file magit-git-executable nil (current-buffer) nil "-c" "alias.echo=!echo" "echo" "x{0}")) (buffer-string)) (and (buffer-name temp-buffer) (kill-buffer temp-buffer))))
  (let ((temp-buffer (generate-new-buffer " *temp*"))) (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn (let ((process-environment (append magit-git-environment process-environment))) (process-file magit-git-executable nil (current-buffer) nil "-c" "alias.echo=!echo" "echo" "x{0}")) (buffer-string)) (and (buffer-name temp-buffer) (kill-buffer temp-buffer)))))
  (equal "x0\n" (let ((temp-buffer (generate-new-buffer " *temp*"))) (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn (let ((process-environment ...)) (process-file magit-git-executable nil (current-buffer) nil "-c" "alias.echo=!echo" "echo" "x{0}")) (buffer-string)) (and (buffer-name temp-buffer) (kill-buffer temp-buffer))))))
  (closure (t) nil (equal "x0\n" (let ((temp-buffer (generate-new-buffer " *temp*"))) (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn (let (...) (process-file magit-git-executable nil ... nil "-c" "alias.echo=!echo" "echo" "x{0}")) (buffer-string)) (and (buffer-name temp-buffer) (kill-buffer temp-buffer)))))))()
  funcall((closure (t) nil (equal "x0\n" (let ((temp-buffer (generate-new-buffer " *temp*"))) (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn (let (...) (process-file magit-git-executable nil ... nil "-c" "alias.echo=!echo" "echo" "x{0}")) (buffer-string)) (and (buffer-name temp-buffer) (kill-buffer temp-buffer))))))))
  eval((funcall (function (closure (t) nil (equal "x0\n" (let ((temp-buffer (generate-new-buffer " *temp*"))) (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn ... ...) (and ... ...)))))))))
  byte-code("\303\304N\211\203
```

It looks like `magit-git-executable`'s default value function assumes that `git -c "alias.exe=!which git | cygpath -wf -" exe` will produce a path to a valid git executable. However, in this situation, it will produce e.g. `C:\cygwin\libexec\git-core\git`, a non-existent path.

This value is later used when calculating the default value of `magit-need-cygwin-noglob`, which attempts to spawn a process with the non-existent path, thus raising the error.

Setting `magit-git-executable` explicitly seems to work around this.

**Comment by npostavs** [comment]
Which Git for Windows version are you using? I would expect `git -c "alias.exe=!which git | cygpath -wf -" exe` should be using Git for Window's `cygpath`. But Git for Windows 1.x doesn't have `cygpath` so we might need something extra to handle that case.

**Comment by CyberShadow** [comment]
Ah, I was indeed using Git for Windows 1.x. I've since updated to 2.x and can no longer reproduce the problem.

I guess this can be closed, seeing as 1.x is obsolete.

**Comment by npostavs** [comment]
Ah, I had already stumbled upon a solution for this when writing the equivalent code for ssh-agency (it works by abusing a bug in Git for Windows 1.x, but that should be fine as it won't see any new releases), and I've been meaning to update that code in magit anyway, so should be fixed for 1.x users now too.

**Comment by tarsius** [comment]
Thanks Noam.

Could you please update the release notes too.

@npostavs @kyleam: I did not add such minor fixes to the release notes myself for the `2.4.0` release, but since I intend to limit myself to bug fixes and very minor changes for the 1-2 months, I think it would be a good idea to do that now. I intend to make a subminor release once a weak.

One minor non-bugfix change I plan to make is the addition of `magit-ref-p` (similar to `magit-branch-p`) which I need in `orgit`. If you have bigger things cooking, then please limit them to feature branches for now. Of course if someone writes a cool new feature, which is ready to be shared, then we can create `2.5` soon - we don't have to wait two months for that. It's just that I do not intend to write any new stuff myself for a little while.

If necessary I'll create a `pu` and/or `maint` branch.
